### PR TITLE
manifest: unref files with range keys when version refcnt is zero

### PR DESCRIPTION
**Pull request by nicktrav**
A manifest version uses reference counting to ensure that obsolete SSTables are only removed once it is safe to do so (i.e. no iterators or snapshots are open referencing tables in a older manifest version).

In #1771, a new slice of `LevelMetadata` was added to support range keys. A slice already existed for point keys (plus range dels). Currently, when a version's ref count reaches zero, only the files present in the point key `LevelMetadata` slice have their ref counts decremented. If a file also contains range keys, the reference count never becomes zero and the file becomes "zombied", and can never be deleted from disk.

Decrement the ref counts on files present in the range keys `LevelMetadata` slice, allowing the ref counts to be correctly zeroed.

Add regression test to exercise the bug observed in cockroachdb/cockroach#90149.

Fix cockroachdb/cockroach#90149.

### CODE REVIEW

**Comment by nicktrav** [comment]
I can probably add some unit tests to the `manifest` package for this case. Thought I'd put this up for review now, with the higher level data-driven test.

**Review by jbowens — COMMENTED** [review]
Right [here in DB.Close](https://github.com/cockroachdb/pebble/blob/5ed983e594da1f186dfac5a082648434d45c2fd3/db.go#L1318) can we error if `len(d.mu.versions.zombieTables) > 0`?

We've already:
1. asserted `Version.Refs() == 0` for all previous versions and `.Refs() == 1` for the current version.
2. and scheduled cleaning jobs to remove all obsolete files, which will remove files from `d.mu.versions.zombieTables`.

I don't think it's possible for `len(d.mu.versions.zombieTables)` to be greater than zero without a ref count leak.
 

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2022#-:-NEqYYd3BdETp6iSLIc_:bf9vjnc)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @itsbilal)

**Review by sumeerbhola — COMMENTED** [review]
Can we assert 0 zombie files and memtables when closing a DB in all our tests, including metamorphic tests?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2022#-:-NEqZK426DhnXQTikqyd:b3vp116)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @itsbilal)

**Review by jbowens — APPROVED** [review]
:lgtm: modulo the `DB.Close` assertion

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2022#-:-NEr6RqtCd05xnqRu43I:b-vc7me4)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @itsbilal)

**Review by jbowens — APPROVED** [review]
:lgtm_strong: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2022#-:-NErC_zN3ImNZoAWHFdt:b-uwkowd)* status: 1 of 4 files reviewed, all discussions resolved (waiting on @itsbilal)

**Review by nicktrav — COMMENTED** [review]
:lgtm:

TFTRs! Ack on the memtable count - let me follow up once this lands.

> in all our tests, including metamorphic tests?

It feels like what we want is some general harness for asserting that a DB and its associated iterators are always closed after tests. Similar to what we have in Cockroach. Let's mull that over on #2032.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2022#-:-NErCbZO01Zq3Qkx3srt:b-cqr0f5)* status: 1 of 4 files reviewed, all discussions resolved (waiting on @itsbilal and @jbowens)

**Pull request merged**
